lib/metag_utils.pm

```diff
@@ -0,0 +1,1795 @@
+    close $fh;
+
+    my $json_data = decode_json($json_text);
+    my $subsys_data = $json_data->{response}->{docs};
+    my $subsys_count = scalar @{$subsys_data};
+    #print $subsys_count;  # 920
+
+    my %subsys_info = ();
+    for (my $i=0; $i<$subsys_count; $i++) {
+        my $subsys_id = $subsys_data->[$i]->{subsystem_id};
+        $subsys_info{$subsys_id}{subsys_class} = $subsys_data->[$i]->{class};
+        $subsys_info{$subsys_id}{subclass} = $subsys_data->[$i]->{subclass};
+        $subsys_info{$subsys_id}{superclass} = $subsys_data->[$i]->{superclass};
+        $subsys_info{$subsys_id}{role_names} = $subsys_data->[$i]->{role_name};
+        $subsys_info{$subsys_id}{role_ids} = $subsys_data->[$i]->{role_id};
+        $subsys_info{$subsys_id}{subsys_name} = $subsys_data->[$i]->{subsystem_name};
+        $subsys_info{$subsys_id}{description} = $subsys_data->[$i]->{description};
+        $subsys_info{$subsys_id}{notes} = $subsys_data->[$i]->{notes};
+    }
+
+    return %subsys_info;
+}
+
+sub _find_function_source {
+    my $self = shift;
+    my $func_tab_ref = shift;
+    my $func = shift;
+
+    my $func_src = 'N/A';
+    foreach my $ftr_id (sort keys %$func_tab_ref) {
+        if (exists($func_tab_ref->{$ftr_id}->{'functions'})) {
+            my $funcs = $func_tab_ref->{$ftr_id}->{'functions'};
+            if ($funcs =~ /\Q$func\E/) {
+                $func_src = $func_tab_ref->{$ftr_id}->{'annotation_src'};
+                last;
+	    }
+        }
+    }
+    return $func_src;
+}
+
+sub _write_html_from_stats {
+    my $self = shift;
+    my $obj_stats_ref = shift;
+    my $gff_stats_ref = shift;
+    my $subsys_ref = shift;
+    my $func_tab_ref = shift;
+    my $template_file = shift;
+    if (ref $obj_stats_ref ne "HASH" || ref $gff_stats_ref ne "HASH"
+             || ref $subsys_ref ne "HASH" || ref $func_tab_ref ne "HASH") {
+        croak "You need to pass in hash references!\n";
+    }
+
+    # dereference the hashes
+    my %obj_stats = %{ $obj_stats_ref };
+    my %gff_stats = %{ $gff_stats_ref };
+    my %subsys_info = %{ $subsys_ref };
+    $template_file = './templates_and_data/table_report.html' unless defined($template_file);
+
+    my $dirname = dirname(__FILE__);
+    my $template = catfile($dirname, $template_file);
+
+    my $fh1 = $self->_openRead($template);
+    read $fh1, my $file_content, -s $fh1; # read the whole file into a string
+    close $fh1;
+
+    my $report_title = "Feature function report for genome <font color=green>$obj_stats{id}</font>";
+    my $rpt_header = "<h3>$report_title:</h3>";
+    my $rpt_data = ("data.addColumn('string', 'function role');\n".
+                    "data.addColumn('string', 'annotation source');\n".
+                    "data.addColumn('number', 'gene count');\n".
+                    "data.addColumn('string', 'subsystem name');\n".
+                    "data.addColumn('string', 'subsystem class');\n".
+                    "data.addRows([\n");
+
```

Comment:
This function would all be better done by outsourcing the work to a template and using Template Toolkit. I can show you how to do this. It's best to separate concerns as much as possible -- keep any HTML generation or editing out of your perl code and in a template file instead.

Reply:
That is a good idea.  I'd love to be able to pass the data to a standalone tool.